airflow/contrib/hooks/gcp_vision_hook.py

```diff
@@ -0,0 +1,254 @@
+            self.log.info('Extracted autogenerated Product ID from the response: %s', product_id)
+
+        return product_id
+
+    @GoogleCloudBaseHook.fallback_to_default_project_id
+    def get_product(self, location, product_id, project_id=None, retry=None, timeout=None, metadata=None):
+        """
+        For the documentation see:
+        :py:class:`~airflow.contrib.operators.gcp_vision_operator.CloudVisionProductGetOperator`
+        """
+        client = self.get_client()
+        name = PRODUCT_NAME_TEMPLATE.format(project_id, location, product_id)
+        self.log.info('Retrieving Product: %s', name)
+        response = self._handle_request(lambda **kwargs: client.get_product(**kwargs), name=name, retry=retry,
+                                        timeout=timeout, metadata=metadata)
+        self.log.info('Product retrieved:\n%s', response)
+        return MessageToJson(response)
+
+    @GoogleCloudBaseHook.fallback_to_default_project_id
+    def update_product(self, product, location=None, product_id=None, update_mask=None, project_id=None,
+                       retry=None, timeout=None, metadata=None):
+        """
+        For the documentation see:
+        :py:class:`~airflow.contrib.operators.gcp_vision_operator.CloudVisionProductUpdateOperator`
+        """
+        client = self.get_client()
+        product = self._determine_product_name(product, product_id, location, project_id)
+        self.log.info('Updating ProductSet: %s', product.name)
+        response = self._handle_request(lambda **kwargs: client.update_product(**kwargs), product=product,
+                                        update_mask=update_mask, retry=retry, timeout=timeout,
+                                        metadata=metadata)
+        self.log.info('Product updated:\n%s', response)
+        return MessageToJson(response)
+
+    @GoogleCloudBaseHook.fallback_to_default_project_id
+    def delete_product(self, location, product_id, project_id=None, retry=None, timeout=None, metadata=None):
+        """
+        For the documentation see:
+        :py:class:`~airflow.contrib.operators.gcp_vision_operator.CloudVisionProductDeleteOperator`
+        """
+        client = self.get_client()
+        name = PRODUCT_NAME_TEMPLATE.format(project_id, location, product_id)
+        self.log.info('Deleting ProductSet: %s', name)
+        response = self._handle_request(lambda **kwargs: client.delete_product(**kwargs), name=name,
+                                        retry=retry, timeout=timeout, metadata=metadata)
+        self.log.info('Product with the name [%s] deleted:', name)
+        return response
+
+    def _handle_request(self, fun, **kwargs):
+        try:
+            return fun(**kwargs)
+        except GoogleAPICallError as e:
+            if isinstance(e, AlreadyExists):
+                raise e
+            else:
+                self.log.error('The request failed, details below:')
```

Comment:
You can not expect logs to also contain an exception. In special cases, it will not be true.

Reply:
Done, I changed it.